pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+func ConcurrencyStateHandler(logger *zap.SugaredLogger, h http.Handler) http.HandlerFunc {
```

Comment:
Please consider adding a benchmark for this when adding the unit tests. Looking at it, it feels like it might add quite a bit of overhead as every request has to be dispatched through the channels and thus through threads.

I wonder if something like this would have the same semantics:

```golang
func ConcurrencyStateHandler(h http.Handler) http.HandlerFunc {
	inFlight := atomic.NewInt(0)
	
 	return func(w http.ResponseWriter, r *http.Request) {'
		if inFlight.Inc() == 1 {
			// Resume
		}
		h.ServeHTTP(w, r)
		if inFlight.Dec() == 0 {
			// Pause
		}
	}
}
```

(spaghetti code, coded in the comment field so likely won't work)

Reply:
We can probably get a rough idea of perf by comparing the handler benchmark with and without this: https://github.com/knative/serving/blob/main/pkg/queue/handler_test.go#L234

Fwiw I think the suggested version with atomic won't quite work because if you're resuming you don't want to serve requests until the resume is complete, which is hard to guarantee with just an atomic (the parallel requests wouldn't block, and I think we need them to to avoid sending to a still-paused container). Similar issues with racing pause and resume calls against each other and in flight requests. 

Anyway fwiw I'd suggest we land the channels version first (I'm pretty sure there are already units tests in the repo we can move over?) and then optimise (unless we could actually do something as simple as this, but I don't think we can) - the code path is optional and the handler won't even be in the chain unless stuff is enabled.

Reply:
Here's a quick comparison of the handler benchmark test with and without the new handler. tl/dr: it's about 3-6x slower with the new handler, ranging 1600-6200 ns/op (vs 475-1700 ns/op before)

Full results

with new handler:
```
goos: linux
goarch: amd64
pkg: knative.dev/serving/pkg/queue
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkProxyHandler/sequential-breaker-10-no-reports-12                 226992              5260 ns/op
BenchmarkProxyHandler/parallel-breaker-10-no-reports-12                   542836              2254 ns/op
BenchmarkProxyHandler/sequential-breaker-infinite-no-reports-12           251680              5165 ns/op
BenchmarkProxyHandler/parallel-breaker-infinite-no-reports-12             636175              1824 ns/op
BenchmarkProxyHandler/sequential-breaker-10-many-reports-12               287947              6193 ns/op
BenchmarkProxyHandler/parallel-breaker-10-many-reports-12                 719083              1670 ns/op
BenchmarkProxyHandler/sequential-breaker-infinite-many-reports-12         311354              5719 ns/op
BenchmarkProxyHandler/parallel-breaker-infinite-many-reports-12           752516              1647 ns/op
PASS
ok      knative.dev/serving/pkg/queue   12.727s
```

without new handler:
```
goos: linux
goarch: amd64
pkg: knative.dev/serving/pkg/queue
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkProxyHandler/sequential-breaker-10-no-reports-12                 714854              1715 ns/op
BenchmarkProxyHandler/parallel-breaker-10-no-reports-12                  2089946               545.5 ns/op
BenchmarkProxyHandler/sequential-breaker-infinite-no-reports-12          1212847              1246 ns/op
BenchmarkProxyHandler/parallel-breaker-infinite-no-reports-12            2435170               474.7 ns/op
BenchmarkProxyHandler/sequential-breaker-10-many-reports-12              1205282              1040 ns/op
BenchmarkProxyHandler/parallel-breaker-10-many-reports-12                2117047               559.2 ns/op
BenchmarkProxyHandler/sequential-breaker-infinite-many-reports-12        1000000              1049 ns/op
BenchmarkProxyHandler/parallel-breaker-infinite-many-reports-12          2402109               488.9 ns/op
PASS
ok      knative.dev/serving/pkg/queue   13.898s
```

Reply:
Can you use benchcmp to give us a nice comparison too?

Reply:
```
benchmark                                                             old ns/op     new ns/op     delta
BenchmarkProxyHandler/sequential-breaker-10-no-reports-12             1715          5260          +206.71%
BenchmarkProxyHandler/parallel-breaker-10-no-reports-12               546           2254          +313.20%
BenchmarkProxyHandler/sequential-breaker-infinite-no-reports-12       1246          5165          +314.53%
BenchmarkProxyHandler/parallel-breaker-infinite-no-reports-12         475           1824          +284.24%
BenchmarkProxyHandler/sequential-breaker-10-many-reports-12           1040          6193          +495.48%
BenchmarkProxyHandler/parallel-breaker-10-many-reports-12             559           1670          +198.64%
BenchmarkProxyHandler/sequential-breaker-infinite-many-reports-12     1049          5719          +445.19%
BenchmarkProxyHandler/parallel-breaker-infinite-many-reports-12       489           1647          +236.88%
```

Reply:
Fwiw I'm sure we can optimise stuff down (and it's always good to do that), but also given each of these is in practice a http request over the network, 5000ns = 0.005ms is probably an acceptable overhead for most use cases, certainly the ones opting in.

Reply:
benchmark added

Reply:
Fair enough, I neglected that this is optional :joy:. We can optimize later for sure, the requirements @julz mentioned make sense and sound like a very nice little concurrency optimization problem :heart: